File: api/controller.js
class Controller { 

    constructor() {
        
        // enum for allowable button strings
        this.buttons = {
            "LEFT_ANALOG": Analog(),
            "RIGHT_ANALOG": Analog(),
            "A_BUTTON": Input(),
            "B_BUTTON": Input(),
            "UP_ARROW": Input(),
            "RIGHT_ARROW": Input(),
            "LEFT_ARROW": Input(),
            "RIGHT_ARROW": Input(),
            "RIGHT_TRIGGER": Input(),
            "LEFT_TRIGGER": Input(),
            "START": Input()
        };
    }

    getState() {
        return Object.assign({}, this.buttons);
    }

    getInput(inputID) {
        if (inputID in self.buttons) {
            return this.buttons.inputID;
        }
    }
}

class Input {

    constructor() {
        this.holdTime = 0;
        this.active = false;
    }

    toggle() {
        this.active = !this.active;
    }

    getHeldTime(){
        return this.heldTime;
    }

    set(active) {
        this.active = active;
    }

    setHoldTime(newHoldTime) {
        this.holdTime = newHoldTime;
    }

}

class Analog extends Input {

    constructor(){
        this.angle = 0;
        this.oomph = 0;
    }

    getAngle() {
        return this.angle;
    }

    getOomph() {
        return this.oomph;
    }
    
    setAngle(angle) {
        this.angle = angle;
    }

    setOopmh (oomph) {
        this.oomph = oomph;
    }
}

File: api/controller.py
from input_types import InputTypes
from input import Input, Analog


class Controller:

    def __init__(self):
        self._buttons = {
            InputTypes.LEFT_ANALOG.value: Analog(),
            InputTypes.RIGHT_ANALOG.value: Analog(),

            InputTypes.LEFT_TRIGGER.value: Analog(),
            InputTypes.RIGHT_TRIGGER.value: Analog(),

            InputTypes.A_BUTTON.value: Input(),
            InputTypes.B_BUTTON.value: Input(),
            InputTypes.Z_BUTTON.value: Input(),

            InputTypes.UP_ARROW.value: Input(),
            InputTypes.LEFT_ARROW.value: Input(),
            InputTypes.RIGHT_ARROW.value: Input(),
            InputTypes.DOWN_ARROW.value: Input(),

            InputTypes.START.value: Input(),
        }

    def get_button(self, button):
        if isinstance(button, InputTypes):
            return self._buttons.get(button.value, None)
        elif isinstance(button, str):
            return self._buttons.get(button, None)

    def get_state(self):
        return {i: s.get_state() for i, s in self._buttons.items()}

